exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java

```diff
@@ -158,38 +165,49 @@ public DrillbitContext getContext() {
     return dContext;
   }
 
-  private ExtendedLatch exitLatch = null; // used to wait to exit when things are still running
-
   /**
    * Waits until it is safe to exit. Blocks until all currently running fragments have completed.
-   *
-   * <p>This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.</p>
+   * This is intended to be used by {@link org.apache.drill.exec.server.Drillbit#close()}.
    */
   public void waitToExit() {
-    synchronized(this) {
-      if (queries.isEmpty() && runningFragments.isEmpty()) {
-        return;
+    final long startTime = System.currentTimeMillis();
+    exitLock.lock();
+
+    try {
+      long diff;
+      while ((diff = (System.currentTimeMillis() - startTime)) < EXIT_TIMEOUT) {
+        if (queries.isEmpty() && runningFragments.isEmpty()) {
+          break;
+        }
+
+        try {
+          final boolean success = exitCondition.await(EXIT_TIMEOUT - diff, TimeUnit.MILLISECONDS);
+
+          if (!success) {
+            break;
+          }
+        } catch (InterruptedException e) {
+          logger.error("Interrupted while waiting to exit");
+        }
       }
 
-      exitLatch = new ExtendedLatch();
-    }
+      if (!(queries.isEmpty() && runningFragments.isEmpty())) {
+        logger.warn("Timed out after %d millis. Shutting down before all fragments and foremen " +
+          "have completed.", EXIT_TIMEOUT);
```

Comment:
The logger has formatting, but does not use `String.format()` formats. Instead, it uses `{}`:

```
logger.warn("Timed out after {} millis...", EXIT_TIMEOUT);
```

Reply:
Done

Reply:
Done